akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,314 @@
+          Actor.same
+
+        case OnSelfUp(subscriber) ⇒
+          if (cluster.selfMember.status == MemberStatus.Up) subscriber ! SelfUp(adaptedCluster.state)
+          else if (!selfSeenUp) {
+            ctx.watch(subscriber)
+            upSubscribers = subscriber :: upSubscribers
+          } else {
+            // self did join, but is now no longer up, we want to avoid subscribing
+            // to not get a memory leak, but also not signal anything
+          }
+          Actor.same
+
+        case OnSelfRemoved(subscriber) ⇒
+          val selfStatus = cluster.selfMember.status
+          if (selfSeenUp && selfStatus == MemberStatus.Removed) subscriber ! SelfRemoved
+          else removedSubscribers = subscriber :: removedSubscribers
+          Actor.same
+
+        case GetCurrentState(sender) ⇒
+          adaptedCluster.sendCurrentClusterState(sender.toUntyped)
+          Actor.same
+
+        case ClusterEvent.MemberUp(member) if member.uniqueAddress == cluster.selfMember.uniqueAddress ⇒
+          selfSeenUp = true
+          upSubscribers.foreach(_ ! SelfUp(cluster.state))
+          upSubscribers = Nil
+          updateCachedSelfMember(member)
+          Actor.same
+
+        case ClusterEvent.MemberRemoved(member, _) if member.uniqueAddress == cluster.selfMember.uniqueAddress ⇒
+          removedSubscribers.foreach(_ ! SelfRemoved)
+          removedSubscribers = Nil
+          updateCachedSelfMember(member)
+          Actor.same
+
+        case m: MemberEvent if m.member.uniqueAddress == cluster.selfMember.uniqueAddress ⇒
+          updateCachedSelfMember(m.member)
+          Actor.same
+
+        case _: MemberEvent ⇒
+          Actor.same
+
+      }
+    }.onSignal {
+
+      case (_, Terminated(ref)) ⇒
+        upSubscribers = upSubscribers.filterNot(_ == ref)
+        removedSubscribers = removedSubscribers.filterNot(_ == ref)
+        Actor.same
+
+    }.narrow[ClusterStateSubscription]
+  }
+
+  private def managerBehavior(adaptedCluster: akka.cluster.Cluster) = Actor.immutable[ClusterCommand]((ctx, msg) ⇒
+    msg match {
+      case Join(address) ⇒
+        adaptedCluster.join(address)
+        Actor.same
+
+      case Leave(address) ⇒
+        adaptedCluster.leave(address)
+        Actor.same
+
+      case Down(address) ⇒
+        adaptedCluster.down(address)
+        Actor.same
+
+      case JoinSeedNodes(addresses) ⇒
+        adaptedCluster.joinSeedNodes(addresses)
+        Actor.same
+
+    }
+
+  )
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] final class AdapterClusterImpl(system: ActorSystem[_]) extends Cluster {
+  import AdapterClusterImpl._
+
+  require(system.isInstanceOf[ActorSystemAdapter[_]], "only adapted actor systems can be used for cluster features")
+  private val untypedSystem = ActorSystemAdapter.toUntyped(system)
+  private def extendedUntyped = untypedSystem.asInstanceOf[ExtendedActorSystem]
+  private val adaptedCluster = akka.cluster.Cluster(untypedSystem)
```

Comment:
name nitpick, perhaps easier to think of it as the "untypedCluster"?

Reply:
I thought so too, but didn't want to nitpick too much ;-)